### PR TITLE
make xrootd proxy, xrootdfs and xrdadler32 work with XrdCl+EC

### PR DESCRIPTION
**Pull request by wyang007**

### CODE REVIEW

**Review by abh3 — CHANGES_REQUESTED** [review]
Several coding style issues (easily correctable) and one significant architectural flaw that is not that easy to fix.

File: src/XrdPosix/XrdPosixXrootd.cc [hunk 1]
```diff
@@ -1060,8 +1060,66 @@ int XrdPosixXrootd::Rename(const char *oldpath, const char *newpath)
 
 // Issue the rename
 //
-  return XrdPosixMap::Result(admin.Xrd.Mv(admin.Url.GetPathWithParams(),
-                             newUrl.GetPathWithParams()));
+   if (! getenv("XRDCL_EC"))
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdPosix/XrdPosixXrootd.cc:1063]: XRDCL_EC is set in XrdCl::PlugInManager::LoadFactory@XrdClPlugInManager.cc. Based on gdb stack trace, this function is called by XrdOssGetSS while loading libXrdPss.so, before XrdPosixConfig::EnvInfo() is called. So we will test XRDCL_EC in XrdPosixConfig::EnvInfo() and XrdPosixGlobals::usingEC there.

File: src/XrdPosix/XrdPosixXrootd.cc [hunk 2]
```diff
@@ -1147,7 +1205,7 @@ int XrdPosixXrootd::Stat(const char *path, struct stat *buf)
 
 // Check if we can get the stat informatation from the cache
 //
-  if (XrdPosixGlobals::theCache)
+  if (! getenv("XRDCL_EC") && XrdPosixGlobals::theCache)
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdPosix/XrdPosixXrootd.cc:1208]: done

File: src/XrdPosix/XrdPosixXrootd.cc [hunk 3]
```diff
@@ -1334,7 +1492,47 @@ int XrdPosixXrootd::Unlink(const char *path)
 
 // Issue the UnLink
 //
-   return XrdPosixMap::Result(admin.Xrd.Rm(admin.Url.GetPathWithParams()));
+   if (! getenv("XRDCL_EC"))
+       return XrdPosixMap::Result(admin.Xrd.Rm(admin.Url.GetPathWithParams()));
+   else
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdPosix/XrdPosixXrootd.cc:1497]: done

File: src/XrdFfs/XrdFfsXrootdfs.cc [hunk 4]
```diff
@@ -799,6 +802,19 @@ static int xrootdfs_read(const char *path, char *buf, size_t size, off_t offset,
 
     fd = (int) fi->fh;
     XrdFfsWcache_flush(fd);  /* in case is the file is reading/writing */
+
+    if (getenv("XRDCL_EC"))
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdFfs/XrdFfsXrootdfs.cc:806]: done

File: src/XrdPosix/XrdPosixXrootd.cc [hunk 5]
```diff
@@ -1334,7 +1492,47 @@ int XrdPosixXrootd::Unlink(const char *path)
 
 // Issue the UnLink
 //
-   return XrdPosixMap::Result(admin.Xrd.Rm(admin.Url.GetPathWithParams()));
+   if (! getenv("XRDCL_EC"))
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdPosix/XrdPosixXrootd.cc:1495]: done

File: src/XrdPosix/XrdPosixXrootd.cc [hunk 6]
```diff
@@ -1060,8 +1060,66 @@ int XrdPosixXrootd::Rename(const char *oldpath, const char *newpath)
 
 // Issue the rename
 //
-  return XrdPosixMap::Result(admin.Xrd.Mv(admin.Url.GetPathWithParams(),
-                             newUrl.GetPathWithParams()));
+   if (! getenv("XRDCL_EC"))
+       return XrdPosixMap::Result(admin.Xrd.Mv(admin.Url.GetPathWithParams(),
+                                  newUrl.GetPathWithParams()));
+   else
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdPosix/XrdPosixXrootd.cc:1066]: This is done for Stat(), Rename() and Unlink()

File: src/XrdHttp/XrdHttpReq.cc [hunk 7]
```diff
@@ -1367,7 +1367,10 @@ int XrdHttpReq::ProcessHTTPReq() {
         l = resourceplusopaque.length() + 1;
         xrdreq.open.dlen = htonl(l);
         xrdreq.open.mode = htons(kXR_ur | kXR_uw | kXR_gw | kXR_gr | kXR_or);
-        xrdreq.open.options = htons(kXR_mkpath | kXR_open_wrto | kXR_delete);
+        if (! getenv("XRDCL_EC"))
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdHttp/XrdHttpReq.cc:1370]: done

File: src/XrdTpc/XrdTpcTPC.cc [hunk 8]
```diff
@@ -729,7 +729,7 @@ int TPCHandler::ProcessPullReq(const std::string &resource, XrdHttpExtReq &req)
     XrdSfsFileOpenMode mode = SFS_O_CREAT;
     auto overwrite_header = req.headers.find("Overwrite");
     if ((overwrite_header == req.headers.end()) || (overwrite_header->second == "T")) {
-        mode = SFS_O_TRUNC;
+        if (! getenv("XRDCL_EC")) mode = SFS_O_TRUNC;
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdTpc/XrdTpcTPC.cc:732]: done

File: src/XrdFfs/XrdFfsXrootdfs.cc [hunk 9]
```diff
@@ -442,7 +442,10 @@ static int xrootdfs_create(const char *path, mode_t mode, struct fuse_file_info
     int res, fd;
     if (!S_ISREG(mode))
         return -EPERM;
-    res = xrootdfs_do_create(path, xrootdfs.rdr, O_CREAT | O_WRONLY, true, &fd);
+    if (getenv("XRDCL_EC"))
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdFfs/XrdFfsXrootdfs.cc:445]: done

File: src/XrdApps/Xrdadler32.cc [hunk 10]
```diff
@@ -237,8 +237,14 @@ int main(int argc, char *argv[])
                 printf("Error_accessing: %s\n", argv[1]);
                 return 1;
             }
-            while ( (len = XrdPosixXrootd::Read(fd, buf, N)) > 0 )
-                adler = adler32(adler, (const Bytef*)buf, len);
+            off_t offset = 0;
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdApps/Xrdadler32.cc:240]: It is a sequential read, so the offset is always the totlen_sofar. In the case of EC, the reading can go beyond the st_size (we agreed that this maybe an unexpected behavior that is best changed in EC). For now, we need to make sure that

1. trim extra part in "len" that went beyond the st_size boundary
2. the next offset doesn't go beyond st_size

File: src/XrdPosix/XrdPosixXrootd.cc [hunk 11]
```diff
@@ -1157,7 +1215,107 @@ int XrdPosixXrootd::Stat(const char *path, struct stat *buf)
 
 // Issue the stat and verify that all went well
 //
-   if (!admin.Stat(*buf)) return -1;
+   if (! getenv("XRDCL_EC"))
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdPosix/XrdPosixXrootd.cc]: done

File: src/XrdPosix/XrdPosixXrootd.cc [hunk 12]
```diff
@@ -1157,7 +1215,107 @@ int XrdPosixXrootd::Stat(const char *path, struct stat *buf)
...
+   {
+       if (!admin.Stat(*buf)) return -1;
+   }
+   else
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdPosix/XrdPosixXrootd.cc]: done

**Review by abh3 — CHANGES_REQUESTED** [review]
New code new set of comments! However, we are quickly converging.

File: src/XrdApps/Xrdadler32.cc [hunk 13]
```diff
@@ -237,8 +237,14 @@ int main(int argc, char *argv[])
                 printf("Error_accessing: %s\n", argv[1]);
                 return 1;
             }
-            while ( (len = XrdPosixXrootd::Read(fd, buf, N)) > 0 )
-                adler = adler32(adler, (const Bytef*)buf, len);
+            off_t totbytes = 0;
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdApps/Xrdadler32.cc:240]: done

File: src/XrdFfs/XrdFfsXrootdfs.cc [hunk 14]
```diff
@@ -89,6 +89,8 @@ static struct fuse_opt xrootdfs_opts[14];
 
 enum { OPT_KEY_HELP, OPT_KEY_SECSSS, };
 
+int usingEC = 0;
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdFfs/XrdFfsXrootdfs.cc:92]: done

File: src/XrdFfs/XrdFfsXrootdfs.cc [hunk 15]
```diff
@@ -127,6 +129,7 @@ static void* xrootdfs_init(struct fuse_conn_info *conn)
     free(pwbuf);
 
 /* put Xrootd related initialization calls here, after fuse daemonize itself. */
+    if (getenv("XRDCL_EC")) usingEC = 1;
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdFfs/XrdFfsXrootdfs.cc:132]: done

File: src/XrdHttp/XrdHttpReq.cc [hunk 16]
```diff
@@ -1367,7 +1368,11 @@ int XrdHttpReq::ProcessHTTPReq() {
         l = resourceplusopaque.length() + 1;
         xrdreq.open.dlen = htonl(l);
         xrdreq.open.mode = htons(kXR_ur | kXR_uw | kXR_gw | kXR_gr | kXR_or);
-        xrdreq.open.options = htons(kXR_mkpath | kXR_open_wrto | kXR_delete);
+        if (usingEC == -1) usingEC = getenv("XRDCL_EC")? 1 : 0;
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdHttp/XrdHttpReq.cc]: done

File: src/XrdTpc/XrdTpcTPC.cc [hunk 17]
```diff
@@ -696,6 +696,7 @@ int TPCHandler::ProcessPushReq(const std::string & resource, XrdHttpExtReq &req)
 #endif
 }
 
+int usingEC = -1;
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdTpc/XrdTpcTPC.cc:699]: done

File: src/XrdTpc/XrdTpcTPC.cc [hunk 18]
```diff
@@ -729,7 +730,8 @@ int TPCHandler::ProcessPullReq(const std::string &resource, XrdHttpExtReq &req)
     XrdSfsFileOpenMode mode = SFS_O_CREAT;
     auto overwrite_header = req.headers.find("Overwrite");
     if ((overwrite_header == req.headers.end()) || (overwrite_header->second == "T")) {
-        mode = SFS_O_TRUNC;
+        if (usingEC == -1) usingEC = getenv("XRDCL_EC")? 1 : 0;
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdTpc/XrdTpcTPC.cc:733]: done

File: src/XrdPosix/XrdPosixXrootd.cc [hunk 19]
```diff
@@ -180,6 +181,12 @@ XrdPosixXrootd::XrdPosixXrootd(int fdnum, int dirnum, int thrnum)
    static XrdSysMutex myMutex;
    char *cfn;
 
+// Test if XRDCL_EC is set. That env var. is set at XrdCl::PlugInManager::LoadFactory
+// in XrdClPlugInManager.cc, which is called (by XrdOssGetSS while loading 
+// libXrdPss.so) before this function. 
+// Note: some standalone programs will call this constructor directly. 
+   if (XrdPosixGlobals::usingEC == -1) XrdPosixGlobals::usingEC = getenv("XRDCL_EC")? 1 : 0;
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdPosix/XrdPosixXrootd.cc:188]: I didn't realize that this is a one-time constructor. that make things much simpler

File: src/XrdPosix/XrdPosixConfig.cc [hunk 20]
```diff
@@ -87,6 +87,7 @@ extern bool                       oidsOK;
 extern bool                       p2lSRC;
 extern bool                       p2lSGI;
 extern bool                       autoPGRD;
+extern int                        usingEC;
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdPosix/XrdPosixConfig.cc]: done

File: src/XrdPosix/XrdPosixConfig.cc [hunk 21]
```diff
@@ -177,6 +178,11 @@ void XrdPosixConfig::EnvInfo(XrdOucEnv &theEnv)
 // method as it picks it up during init time. We leave the code for historical
 // reasons but we really should have gotten rid of EnvInfo()!
 // if (XrdPosixGlobals::myCache2) XrdPosixGlobals::myCache2->EnvInfo(theEnv);
+
+// Test if XRDCL_EC is set. That env var. is set at XrdCl::PlugInManager::LoadFactory
+// in XrdClPlugInManager.cc, which is called (by XrdOssGetSS while loading 
+// libXrdPss.so) before this function. 
+   XrdPosixGlobals::usingEC = getenv("XRDCL_EC")? 1 : 0;
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdPosix/XrdPosixConfig.cc]: done

File: src/XrdPosix/XrdPosixXrootd.cc [hunk 22]
```diff
@@ -94,6 +94,7 @@ bool             oidsOK    = false;
 bool             p2lSRC    = false;
 bool             p2lSGI    = false;
 bool             autoPGRD  = false;
+int              usingEC   = -1;
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdPosix/XrdPosixXrootd.cc]: done

File: src/XrdPosix/XrdPosixXrootd.cc (hunk 19 above)
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdPosix/XrdPosixXrootd.cc]: done

File: src/XrdPosix/XrdPosixXrootd.cc [hunk 23]
```diff
@@ -1040,6 +1047,7 @@ int XrdPosixXrootd::Readdir64_r(DIR *dirp, struct dirent64  *entry,
 /*                                R e n a m e                                 */
 /******************************************************************************/
 
+int ec_rename(const char*, const char*, XrdPosixAdmin*);
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdPosix/XrdPosixXrootd.cc]: I moved they as private methods of the class, as you wish. I just didn't think the anyone who read the .hh needs to even know the existence of these methods, as they serve no other purposed than making the code more readable.

done

File: src/XrdPosix/XrdPosixXrootd.cc [hunk 24]
```diff
@@ -1133,6 +1144,7 @@ void XrdPosixXrootd::Seekdir(DIR *dirp, long loc)
 /*                                  S t a t                                   */
 /******************************************************************************/
   
+int ec_stat(const char*, struct stat*, XrdPosixAdmin*);
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdPosix/XrdPosixXrootd.cc]: done

File: src/XrdPosix/XrdPosixXrootd.cc [hunk 25]
```diff
@@ -1316,6 +1331,7 @@ int XrdPosixXrootd::Truncate(const char *path, off_t Size)
 /*                                 U n l i n k                                */
 /******************************************************************************/
 
+int ec_unlink(const char*, XrdPosixAdmin*);
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdPosix/XrdPosixXrootd.cc]: done

File: src/XrdPosix/XrdPosixXrootd.cc [hunk 26]
```diff
@@ -1464,3 +1483,213 @@ int XrdPosixXrootd::Fault(XrdPosixFile *fp, int ecode)
    errno = ecode;
    return -1;
 }
+
```
[the remark this answers is not in the document]
Comment (wyang007) [reply src/XrdPosix/XrdPosixXrootd.cc:1483]: done

File: src/XrdPosix/XrdPosixXrootd.hh [hunk 27]
```diff
@@ -46,6 +46,7 @@
 
 #include "XrdPosix/XrdPosixOsDep.hh"
 #include "XrdSys/XrdSysPthread.hh"
+#include "XrdPosix/XrdPosixAdmin.hh"
```
Comment (abh3) [inline src/XrdPosix/XrdPosixXrootd.hh:49]: I'd prefer you forward declare this object. No need to mess up the where PosixXrootd.hh is included as only needed in the cc file.

File: src/XrdPosix/XrdPosixXrootd.hh [hunk 28]
```diff
@@ -46,6 +46,7 @@
...
 
 struct XrdOucIOVec;
```
Comment (abh3) [inline src/XrdPosix/XrdPosixXrootd.hh:51]: And here is a good place to forward declare XrdPosixAdmin

**Review by abh3 — CHANGES_REQUESTED** [review]

File: src/XrdPosix/XrdPosixXrootd.hh (hunk 27 above)
Comment (wyang007) [reply src/XrdPosix/XrdPosixXrootd.hh]: done

File: src/XrdPosix/XrdPosixXrootd.hh (hunk 28 above)
Comment (wyang007) [reply src/XrdPosix/XrdPosixXrootd.hh]: done

**Pull request merged**